frontend/javascripts/oxalis/model/sagas/isosurface_saga.ts

```diff
@@ -506,6 +515,66 @@ function* _refreshIsosurfaceWithMap(
  * Precomputed Meshes
  *
  */
+
+// Avoid redundant fetches of mesh files for the same layer by
+// storing Deferreds per layer lazily.
+const fetchDeferredsPerLayer: Record<string, Deferred<Array<APIMeshFile>, unknown>> = {};
+function* maybeFetchMeshFiles(action: MaybeFetchMeshFilesAction): Saga<void> {
+  const { segmentationLayer, dataset, mustRequest, autoActivate, callback } = action;
+
+  if (!segmentationLayer) {
+    callback([]);
+    return;
+  }
+
+  const layerName = segmentationLayer.name;
+
+  function* maybeActivateMeshFile(availableMeshFiles: APIMeshFile[]) {
+    const currentMeshFile = yield* select(
+      (state) => state.localSegmentationData[layerName].currentMeshFile,
+    );
+    if (!currentMeshFile && availableMeshFiles.length > 0 && autoActivate) {
+      yield* put(updateCurrentMeshFileAction(layerName, availableMeshFiles[0].meshFileName));
+    }
+  }
+
+  // If a deferred already exists, the one can be awaited (regardless of
+  // whether it's finished or not) and its content used to call the callback.
+  // If mustRequest was set to true, a new deferred will be created which
+  // replaces the old one (old references to the first Deferred will still
+  // work and will be resolved by the corresponding saga execution).
+  if (fetchDeferredsPerLayer[layerName] && !mustRequest) {
+    const meshes = yield* call(() => fetchDeferredsPerLayer[layerName].promise());
```

Comment:
TS doesn't complain, but the `this` context is lost in that case. I could change the methods of the Deferred class to use something like `resolve = () => {}` instead of `resolve() {}`, but if we go down this route, we would have to do this everywhere in the long run, which feels weird :thinking: Not too sure about this. Feel free to disagree :)

Reply:
Ah, makes sense, let's leave it as-is then :+1: